elements/lisk-chain/src/block_header.ts

```diff
@@ -0,0 +1,286 @@
+/*
+ * Copyright © 2021 Lisk Foundation
+ *
+ * See the LICENSE file at the top-level directory of this distribution
+ * for licensing information.
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+import { objects } from '@liskhq/lisk-utils';
+import { signDataWithPrivateKey, hash } from '@liskhq/lisk-cryptography';
+import { codec } from '@liskhq/lisk-codec';
+import { validator, LiskValidationError } from '@liskhq/lisk-validator';
+import { TAG_BLOCK_HEADER } from './constants';
+
+export const signingBlockHeaderSchema = {
+	$id: '/block/header/signing/2',
+	type: 'object',
+	properties: {
+		version: { dataType: 'uint32', fieldNumber: 1 },
+		timestamp: { dataType: 'uint32', fieldNumber: 2 },
+		height: { dataType: 'uint32', fieldNumber: 3 },
+		previousBlockID: { dataType: 'bytes', fieldNumber: 4 },
+		stateRoot: { dataType: 'bytes', fieldNumber: 5 },
+		transactionRoot: { dataType: 'bytes', fieldNumber: 6 },
+		generatorAddress: { dataType: 'bytes', fieldNumber: 7 },
+		assets: {
+			type: 'array',
+			fieldNumber: 8,
+			items: {
+				type: 'object',
+				required: ['moduleID', 'data'],
+				properties: {
+					moduleID: { dataType: 'uint32', fieldNumber: 1 },
+					data: { dataType: 'bytes', fieldNumber: 2 },
+				},
+			},
+		},
+	},
+	required: [
+		'version',
+		'timestamp',
+		'height',
+		'previousBlockID',
+		'stateRoot',
+		'transactionRoot',
+		'generatorAddress',
+		'assets',
+	],
+};
+
+export const blockHeaderSchema = {
+	...signingBlockHeaderSchema,
+	$id: '/block/header/2',
+	required: [...signingBlockHeaderSchema.required, 'signature'],
+	properties: {
+		...signingBlockHeaderSchema.properties,
+		signature: { dataType: 'bytes', fieldNumber: 9 },
+	},
+};
+
+export const blockHeaderSchemaWithId = {
+	...blockHeaderSchema,
+	$id: '/block/header/2',
+	required: [...blockHeaderSchema.required, 'id'],
+	properties: {
+		...blockHeaderSchema.properties,
+		id: { dataType: 'bytes', fieldNumber: 10 },
+	},
+};
+
+interface BlockHeaderAsset {
+	moduleID: number;
+	data: Buffer;
+}
+
+interface BlockHeaderAttrs {
+	readonly version: number;
+	readonly height: number;
+	readonly generatorAddress: Buffer;
+	readonly previousBlockID: Buffer;
+	readonly timestamp?: number;
+	readonly stateRoot?: Buffer;
+	readonly transactionRoot?: Buffer;
+	readonly assets: ReadonlyArray<BlockHeaderAsset>;
+	signature?: Buffer;
+	id?: Buffer;
+}
+
+export interface BlockHeaderJSON {
```

Comment:
if we are exporting then maybe move it to `types.ts`

Reply:
Same as above.